File: sdks/dotnet/src/PrismDb/Codec.cs
// Low-level binary codec: a growable little-endian Writer, a bounds-checked
// Reader, and the length-prefixed frame helpers. The byte layouts mirror
// crates/prism-protocol/src/codec.rs exactly (all multi-byte integers LE).

using System;
using System.Text;

namespace PrismDb
{
    /// <summary>A growable little-endian writer over a byte buffer.</summary>
    public sealed class Writer
    {
        private byte[] _buf;
        private int _len;

        public Writer(int capacity = 64)
        {
            _buf = new byte[capacity < 8 ? 8 : capacity];
            _len = 0;
        }

        private void Ensure(int extra)
        {
            int needed = _len + extra;
            if (needed <= _buf.Length) return;
            int cap = _buf.Length * 2;
            while (cap < needed) cap *= 2;
            var next = new byte[cap];
            Array.Copy(_buf, next, _len);
            _buf = next;
        }

        public void U8(int v)
        {
            Ensure(1);
            _buf[_len++] = (byte)(v & 0xFF);
        }

        public void U16(int v)
        {
            Ensure(2);
            _buf[_len++] = (byte)(v & 0xFF);
            _buf[_len++] = (byte)((v >> 8) & 0xFF);
        }

        public void U32(long v)
        {
            Ensure(4);
            _buf[_len++] = (byte)(v & 0xFF);
            _buf[_len++] = (byte)((v >> 8) & 0xFF);
            _buf[_len++] = (byte)((v >> 16) & 0xFF);
            _buf[_len++] = (byte)((v >> 24) & 0xFF);
        }

        public void I32(int v) => U32(unchecked((uint)v));

        public void U64(ulong v)
        {
            Ensure(8);
            for (int i = 0; i < 8; i++) _buf[_len++] = (byte)((v >> (i * 8)) & 0xFF);
        }

        public void I64(long v) => U64(unchecked((ulong)v));

        public void F64(double v) => U64(unchecked((ulong)BitConverter.DoubleToInt64Bits(v)));

        /// <summary>A 128-bit unsigned integer as 16 little-endian bytes.</summary>
        public void U128(ulong lo, ulong hi)
        {
            U64(lo);
            U64(hi);
        }

        public void Raw(byte[] bytes)
        {
            Ensure(bytes.Length);
            Array.Copy(bytes, 0, _buf, _len, bytes.Length);
            _len += bytes.Length;
        }

        /// <summary>A UTF-8 string with a u16 length prefix.</summary>
        public void StrU16(string s)
        {
            var b = Encoding.UTF8.GetBytes(s);
            U16(b.Length);
            Raw(b);
        }

        /// <summary>A UTF-8 string with a u32 length prefix.</summary>
        public void StrU32(string s)
        {
            var b = Encoding.UTF8.GetBytes(s);
            U32(b.Length);
            Raw(b);
        }

        /// <summary>A byte string with a u16 length prefix.</summary>
        public void BytesU16(byte[] b)
        {
            U16(b.Length);
            Raw(b);
        }

        /// <summary>A byte string with a u32 length prefix.</summary>
        public void BytesU32(byte[] b)
        {
            U32(b.Length);
            Raw(b);
        }

        /// <summary>The written bytes (a fresh copy).</summary>
        public byte[] Out()
        {
            var o = new byte[_len];
            Array.Copy(_buf, o, _len);
            return o;
        }
    }

    /// <summary>A bounds-checked little-endian reader over a byte array.</summary>
    public sealed class Reader
    {
        private readonly byte[] _buf;
        private int _p;

        public Reader(byte[] buf)
        {
            _buf = buf;
            _p = 0;
        }

        private void Need(int n)
        {
            if (_p + n > _buf.Length)
                throw new ProtocolException($"truncated: need {n} bytes at offset {_p}");
        }

        public int U8()
        {
            Need(1);
            return _buf[_p++];
        }

        public int U16()
        {
            Need(2);
            int v = _buf[_p] | (_buf[_p + 1] << 8);
            _p += 2;
            return v;
        }

        public long U32()
        {
            Need(4);
            long v = (uint)(_buf[_p] | (_buf[_p + 1] << 8) | (_buf[_p + 2] << 16) | (_buf[_p + 3] << 24));
            _p += 4;
            return v;
        }

        public int I32()
        {
            Need(4);
            int v = _buf[_p] | (_buf[_p + 1] << 8) | (_buf[_p + 2] << 16) | (_buf[_p + 3] << 24);
            _p += 4;
            return v;
        }

        public ulong U64()
        {
            Need(8);
            ulong v = 0;
            for (int i = 0; i < 8; i++) v |= (ulong)_buf[_p + i] << (i * 8);
            _p += 8;
            return v;
        }

        public long I64() => unchecked((long)U64());

        public double F64() => BitConverter.Int64BitsToDouble(unchecked((long)U64()));

        /// <summary>Returns the (lo, hi) 64-bit halves of a 128-bit value.</summary>
        public (ulong Lo, ulong Hi) U128()
        {
            ulong lo = U64();
            ulong hi = U64();
            return (lo, hi);
        }

        public byte[] Raw(int n)
        {
            Need(n);
            var s = new byte[n];
            Array.Copy(_buf, _p, s, 0, n);
            _p += n;
            return s;
        }

        public string StrU16() => Encoding.UTF8.GetString(Raw(U16()));

        public string StrU32() => Encoding.UTF8.GetString(Raw((int)U32()));

        public byte[] BytesU16() => Raw(U16());

        public byte[] BytesU32() => Raw((int)U32());

        public int Remaining => _buf.Length - _p;

        /// <summary>Throw unless every byte has been consumed.</summary>
        public void ExpectEnd()
        {
            if (Remaining != 0)
                throw new ProtocolException($"{Remaining} trailing byte(s) after message");
        }
    }

    public static class Frame
    {
        /// <summary>Wrap a payload in a [len:u32][payload] frame.</summary>
        public static byte[] Encode(byte[] payload)
        {
            var o = new byte[4 + payload.Length];
            uint len = (uint)payload.Length;
            o[0] = (byte)(len & 0xFF);
            o[1] = (byte)((len >> 8) & 0xFF);
            o[2] = (byte)((len >> 16) & 0xFF);
            o[3] = (byte)((len >> 24) & 0xFF);
            Array.Copy(payload, 0, o, 4, payload.Length);
            return o;
        }
    }
}

File: sdks/python/prismdb/_codec.py
"""Low-level binary codec: a growable little-endian ``Writer``, a bounds-checked
``Reader``, and the length-prefixed frame helpers. The byte layouts mirror
``crates/prism-protocol/src/codec.rs`` exactly (all multi-byte integers LE)."""

from __future__ import annotations

import struct

from .errors import ProtocolError

_U64_MASK = (1 << 64) - 1
_U128_MASK = (1 << 128) - 1


class Writer:
    """A growable little-endian writer over a bytearray."""

    __slots__ = ("_buf",)

    def __init__(self) -> None:
        self._buf = bytearray()

    def u8(self, v: int) -> None:
        self._buf.append(v & 0xFF)

    def u16(self, v: int) -> None:
        self._buf += struct.pack("<H", v & 0xFFFF)

    def u32(self, v: int) -> None:
        self._buf += struct.pack("<I", v & 0xFFFFFFFF)

    def i32(self, v: int) -> None:
        self._buf += struct.pack("<i", _to_signed(v, 32))

    def u64(self, v: int) -> None:
        self._buf += struct.pack("<Q", v & _U64_MASK)

    def i64(self, v: int) -> None:
        self._buf += struct.pack("<q", _to_signed(v, 64))

    def f64(self, v: float) -> None:
        self._buf += struct.pack("<d", v)

    def u128(self, v: int) -> None:
        """A 128-bit unsigned integer as 16 little-endian bytes."""
        x = v & _U128_MASK
        self._buf += struct.pack("<QQ", x & _U64_MASK, (x >> 64) & _U64_MASK)

    def raw(self, b: bytes) -> None:
        self._buf += b

    def str_u16(self, s: str) -> None:
        """A UTF-8 string with a u16 length prefix."""
        b = s.encode("utf-8")
        self.u16(len(b))
        self._buf += b

    def str_u32(self, s: str) -> None:
        """A UTF-8 string with a u32 length prefix."""
        b = s.encode("utf-8")
        self.u32(len(b))
        self._buf += b

    def bytes_u16(self, b: bytes) -> None:
        """A byte string with a u16 length prefix."""
        self.u16(len(b))
        self._buf += b

    def bytes_u32(self, b: bytes) -> None:
        """A byte string with a u32 length prefix."""
        self.u32(len(b))
        self._buf += b

    def out(self) -> bytes:
        return bytes(self._buf)


class Reader:
    """A bounds-checked little-endian reader over a bytes-like object."""

    __slots__ = ("_buf", "_p")

    def __init__(self, buf: bytes) -> None:
        self._buf = buf
        self._p = 0

    def _need(self, n: int) -> None:
        if self._p + n > len(self._buf):
            raise ProtocolError(f"truncated: need {n} bytes at offset {self._p}")

    def u8(self) -> int:
        self._need(1)
        v = self._buf[self._p]
        self._p += 1
        return v

    def u16(self) -> int:
        self._need(2)
        v = struct.unpack_from("<H", self._buf, self._p)[0]
        self._p += 2
        return v

    def u32(self) -> int:
        self._need(4)
        v = struct.unpack_from("<I", self._buf, self._p)[0]
        self._p += 4
        return v

    def i32(self) -> int:
        self._need(4)
        v = struct.unpack_from("<i", self._buf, self._p)[0]
        self._p += 4
        return v

    def u64(self) -> int:
        self._need(8)
        v = struct.unpack_from("<Q", self._buf, self._p)[0]
        self._p += 8
        return v

    def i64(self) -> int:
        self._need(8)
        v = struct.unpack_from("<q", self._buf, self._p)[0]
        self._p += 8
        return v

    def f64(self) -> float:
        self._need(8)
        v = struct.unpack_from("<d", self._buf, self._p)[0]
        self._p += 8
        return v

    def u128(self) -> int:
        self._need(16)
        lo, hi = struct.unpack_from("<QQ", self._buf, self._p)
        self._p += 16
        return (hi << 64) | lo

    def raw(self, n: int) -> bytes:
        self._need(n)
        s = self._buf[self._p : self._p + n]
        self._p += n
        return bytes(s)

    def str_u16(self) -> str:
        return self.raw(self.u16()).decode("utf-8")

    def str_u32(self) -> str:
        return self.raw(self.u32()).decode("utf-8")

    def bytes_u16(self) -> bytes:
        return self.raw(self.u16())

    def bytes_u32(self) -> bytes:
        return self.raw(self.u32())

    def remaining(self) -> int:
        return len(self._buf) - self._p

    def expect_end(self) -> None:
        """Raise unless every byte has been consumed."""
        if self.remaining() != 0:
            raise ProtocolError(f"{self.remaining()} trailing byte(s) after message")


def _to_signed(v: int, bits: int) -> int:
    """Wrap ``v`` into a signed ``bits``-wide integer (two's complement)."""
    mask = (1 << bits) - 1
    v &= mask
    if v >= 1 << (bits - 1):
        v -= 1 << bits
    return v


def frame_encode(payload: bytes) -> bytes:
    """Wrap a payload in a ``[len:u32][payload]`` frame."""
    return struct.pack("<I", len(payload)) + payload
